cuda/src/memory/cuda/device_memory_resource.cpp

```diff
@@ -0,0 +1,70 @@
+/*
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include "../../utils/cuda_error_handling.hpp"
+#include "../../utils/cuda_wrappers.hpp"
+
+#include "vecmem/memory/resources/memory_resource.hpp"
+#include "vecmem/memory/cuda/device_memory_resource.hpp"
+
+#include <cuda_runtime_api.h>
+
+namespace vecmem::cuda {
+    device_memory_resource::device_memory_resource(
+        int device
+    ) :
+        m_device(device >= 0 ? device : details::get_device())
+    {
+    }
+
+    void * device_memory_resource::do_allocate(
+        std::size_t bytes,
+        std::size_t
+    ) {
+        int current_device = 0;
+
+        VECMEM_CUDA_ERROR_CHECK(cudaGetDevice(&current_device));
+        VECMEM_CUDA_ERROR_CHECK(cudaSetDevice(m_device));
+
+        void * res;
+        VECMEM_CUDA_ERROR_CHECK(cudaMalloc(&res, bytes));
+
+        VECMEM_CUDA_ERROR_CHECK(cudaSetDevice(current_device));
+
+        return res;
+    }
+
+    void device_memory_resource::do_deallocate(
+        void * p,
+        std::size_t,
+        std::size_t
+    ) {
+        int current_device = 0;
+
+        VECMEM_CUDA_ERROR_CHECK(cudaGetDevice(&current_device));
+        VECMEM_CUDA_ERROR_CHECK(cudaSetDevice(m_device));
+
+        VECMEM_CUDA_ERROR_CHECK(cudaFree(p));
+
+        VECMEM_CUDA_ERROR_CHECK(cudaSetDevice(current_device));
+    }
```

Comment:
I think `cudaFree` does not need all this machinery with switching the current device around.

But...

Since you introduced `cuda_wrappers.hpp` / `cuda_wrappers.cpp` anyway, I wonder if it would be possible to introduce a helper construct for these sort of calls. I would imagine the following kind of UI:

```c++
details::run_on_device( m_device )( []{
      VECMEM_CUDA_ERROR_CHECK( cudaFree( p ) );
   } );
```

Naturally `vecmem::cuda::details::run_on_device` would be (sort of) a functor, that could execute "something executable" while setting the current device to the value that it received in its constructor.

Such a construct could come in handy in some other places as well later on...

Reply:
Hmm, that's an interesting idea. I was actually thinking of an RAII-style device selection mechanism similar to `std::lock_guard` earlier, where all the setting and restoring would happen in the constructor and destructor of an object, respectively:

```cpp
void * device_memory_resource::do_allocate(
    std::size_t bytes,
    std::size_t
) {
    details::select_device(m_device);

    void * res;
    VECMEM_CUDA_ERROR_CHECK(cudaMalloc(&res, bytes));

    return res;
}
```

I think these are both elegant ways to handle the problem at hand here.

Reply:
:thinking: You're right, that would be a little easier to implement. And I myself have also used that sort of code for handling "ROOT directories" in my analysis code in the past. I just didn't think of it in this instance...

Reply:
If you're happy with the code as it is now, I'd suggest we merge this and I can start to work on a nicer way to manage the device selection. :smile: